bob-common/src/mapper.rs

```diff
@@ -131,18 +125,39 @@ impl Virtual {
         trace!("extract indexes of target nodes");
         trace!("nodes available: {}", self.nodes.len());
         let offset = self.support_nodes_offset.fetch_add(1, Ordering::Relaxed);
-        let mut support_nodes = Vec::with_capacity(count);
-        for (id, node) in self.nodes.iter().skip(offset % self.nodes.len()) {
-            if target_nodes.iter().all(|i| i.index() != *id) {
-                support_nodes.push(node);
+        let mut support_nodes: Vec<&Node> = Vec::with_capacity(count);
+        let mut available_count = 0;
+        let mut offset_counter = offset % self.nodes.len();
+        let mut iterator = self.nodes.iter().enumerate();
```

Comment:
No need to use iterator with vector. It is much simpler to use indexation: `self.nodes[i]`. Thus `for i in 0..nodes.len()` or some kind of `while` loop will be better

Reply:
Tested algo on C#:
```C#
    private (int index, int avail) Find(bool[] status, int offset)
    {
        int avail = 0;
        for (int i = 0; i < status.Length; i++)
        {
            if (status[i])
            {
                if (avail == offset)
                    return (i, avail + 1);

                avail++;
            }
        }

        return (-1, avail);
    }

    public int Select(bool[] status)
    {
        int offset = Interlocked.Increment(ref _offset);
        var res = Find(status, offset % status.Length);

        if (res.index < 0 && res.avail > 0)
        {
            int curr_offset = offset % res.avail;
            while (res.index < 0 && res.avail > 0)
            {
                res = Find(status, curr_offset);
                curr_offset = Math.Max(0, curr_offset - res.avail);
            }
        }

        if (res.index < 0)
            return offset % status.Length;

        return res.index;
    }
```